client/assets-proof/src/lib.rs

```diff
@@ -0,0 +1,691 @@
+            .collect::<BTreeMap<_, _>>();
+
+        Ok(AssetsProof {
+            version: target_version,
+            timestamp_usecs,
+            currencies,
+            total_unfrozen_balances,
+            parent_vasp,
+            child_vasps,
+        })
+    }
+
+    /// Resolve the current chain metadata according the the config.
+    ///
+    /// 1. If the version is pre-specified, then we get the metadata for that
+    ///    specific version.
+    /// 2. If a timestamp is pre-specified, then get the metadata for the last
+    ///    state at or before that timestamp.
+    /// 3. Otherwise, we just get the latest metadata.
+    fn resolve_metadata(&self, client: &impl Client) -> Result<Response<MetadataView>> {
+        if let Some(version) = self.version {
+            client.get_metadata_by_version(version)
+        } else if let Some(timestamp_usecs) = self.timestamp_usecs {
+            let current_metadata = client.get_metadata()?;
+            let latest_version = current_metadata.state().version;
+            let past_version =
+                client.get_last_version_before_timestamp(timestamp_usecs, latest_version)?;
+            client.get_metadata_by_version(past_version)
+        } else {
+            client.get_metadata()
+        }
+    }
+}
+
+pub trait Client {
+    fn get_last_version_before_timestamp(
+        &self,
+        timestamp_usecs: u64,
+        version: Version,
+    ) -> Result<Version>;
+
+    fn get_metadata(&self) -> Result<Response<MetadataView>>;
+
+    fn get_metadata_by_version(&self, version: Version) -> Result<Response<MetadataView>>;
+
+    fn get_currencies(&self) -> Result<Response<Vec<CurrencyInfoView>>>;
+
+    fn get_account_by_version(
+        &self,
+        address: AccountAddress,
+        version: Version,
+        currency_ids: &[Identifier],
+    ) -> Result<Response<Option<AccountView>>>;
+}
+
+impl Client for diem_client::BlockingClient {
+    fn get_last_version_before_timestamp(
+        &self,
+        _timestamp_usecs: u64,
+        _version: Version,
+    ) -> Result<Version> {
+        todo!()
+    }
+
+    fn get_metadata(&self) -> Result<Response<MetadataView>> {
+        self.get_metadata().context("Failed to get ledger metadata")
+    }
+
+    fn get_metadata_by_version(&self, version: Version) -> Result<Response<MetadataView>> {
+        self.get_metadata_by_version(version)
+            .with_context(|| format!("Failed to get ledger metadata: version={}", version))
+    }
+
+    fn get_currencies(&self) -> Result<Response<Vec<CurrencyInfoView>>> {
+        self.get_currencies()
+            .context("Failed to get current supported ledger currencies")
+    }
+
+    fn get_account_by_version(
+        &self,
+        address: AccountAddress,
+        version: Version,
+        currency_ids: &[Identifier],
+    ) -> Result<Response<Option<AccountView>>> {
+        // Hack work around for get_account_by_version via get_account_state_with_proof,
```

Comment:
Yep, I've improved the description here. It's not as unreasonable as it sounds, since we just get the raw account state and throw out the state proof.